include/storage/fbw.h

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright (c) 2017, 2020 Nordic Semiconductor ASA
+ * Copyright (c) 2017 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Public API for flash buffered writes
+ */
+
+#ifndef ZEPHYR_INCLUDE_STORAGE_FBW_H_
+#define ZEPHYR_INCLUDE_STORAGE_FBW_H_
+
+/**
+ * @brief Abstraction over flash buffered writes.
+ *
+ * @defgroup fbw flash buffered write Interface
+ * @{
+ */
+
+#include <stdbool.h>
+#include <drivers/flash.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef int (*fbw_callback_t)(u8_t *buf, size_t len, size_t offset);
+
+/**
+ * @brief Structure for flash buffered writes
+ *
+ * Users should treat these structures as opaque values and only interact
+ * with them through the below API.
+ */
+struct fbw_ctx {
+	u8_t *buf; /* Write buffer */
+	size_t buf_len; /* Length of write buffer */
+	size_t buf_bytes; /* Number of bytes currently stored in write buf */
+	struct device *fdev; /* Flash device */
+	size_t bytes_written; /* Number of bytes written to flash */
+	size_t offset; /* Offset from base of flash device to write area */
+	size_t available; /* Available bytes in write area */
+	fbw_callback_t callback; /* Callback invoked on completed write op */
+#ifdef CONFIG_FBW_ERASE
+	off_t off_last; /* Last erased offset */
+#endif
+};
+
+/**
+ * @brief Initialize context needed for buffered write to flash.
+ *
+ * @param ctx context to be initialized
+ * @param fdev Flash device to operate on
+ * @param buf Write buffer
+ * @param buf_len Length of write buffer
+ * @param offset Offset within flash device
+ * @param size Size of region within flash device
```

Comment:
Looking at this with fresh eyes, does this mean the maximum number of bytes to write is `size - offset`? Or is it `size`?

Might want to clarify in the API docs here so lazy folks like me don't have to read the code ;)

Reply:
In my original version I was providing the `end` address instead of the size. Whilst making it more clear what the user should expect, I think it is more convenient to use the size, as this information is typically more available (start - size of partitions is more common info than start - end address).

Also I'm not sure if the `send 0 to get the total size` functionality is really that useful, so feedback on that is welcome.

Reply:
> Also I'm not sure if the `send 0 to get the total size` functionality is really that useful, so feedback on that is welcome.

+1 to this and the change to make it the size of the buffer area